# Problem set 1

All work should be done in this notebook. Upload the `.ipynb` file to CMS.

## Part 1: Birthday collisions 

We will work on this problem together in the Friday session.

We often use probabilistic models to decide whether events that seem surprising are actually unlikely. The birthday problem is a classic example of such a phenomenon:  What is the probability that at least two people in a group of size `n` have the same birthday? Assuming that each year has 365 days (no leap years) and that each day is equally likely as a birthdate, __write a function which returns this probability as a function of the group size `n`__.

As a hint, notice that enumerating all of the possible ways in which a group of individuals might share at least one birthday is quite challenging.  There could be one overlapping birthday, two overlapping birthdays, and so on. Instead, we'll calculate the probability that *no one* shares a birthday, and then subtract that value from 1.0.

In [ ]:
##Brainstorming
 n_distinct = 365 - (n + 1)
 n_birthday = n * 365
 no_one_shares = n_distinct / n_birthday

 return 1 - no_one_shares

In [3]:
##answer here
def compute_birthday_prob(n):
 ## return the probability that there exists atleast 2 people in a class of size n who share the same birthday   
 p = 1 ## the probability that no 2 individuals in a class share a birthday
 for number_of_free_birthdays in range(365-n+1,366):
    p = p * number_of_free_birthdays
 return 1 - p / (365 ** n)


Using your function, print the probability that at least 2 people in a class of size $n=23$ share a birthday.

In [4]:
##answer here
n = 23
print(compute_birthday_prob(n))

0.5072972343239854


Now, let's write a second program to verify our theoretical results via random sampling.  Write a function which simulates the birthday problem for a given `num_trials` number of trials for a group of size `n` (*HINT: np.random.randint and Counters*).

In [ ]:
##Brainstorming
n = 
def simulator(n):
    for student in range(n):
        student = np.random.randint 


np.random.randint 

In [9]:
##Hint: Counters have built in method.most_common() that you should use
###answer here
from numpy.random import randint
from collections import Counter
def generate_random_brithday_sample(n):
    ##assign random birthdays to n people
    birthdays = randint(1,366,n)
    ## dictionary with key values keys = brithday values = # of ppl w/ birthday
    birthday_counts = Counter(birthdays) 
    birthday_counts = birthday_counts.most_common()

    if birthday_counts[0][1] > 1:
        ##duplicate birthday
        return 1
    else: 
        ##No duplicates
        return 0


In [10]:
print(generate_random_brithday_sample(50))

0


Using the function you have just written, empirically estimate the probability that at least two individuals in a group of size `n` = 23 share a birthday using `num_trials` = 10000 samples

In [14]:
##answer here
def estimate_birthday_probability(num_trials,n):
    ## return an estiamte for the prob that there is atleast 1 overlapping bday in a group of size n
    p = 0 #initialization of probability estimate
    for _ in range(num_trials):
        p = p + generate_random_brithday_sample(n)
    return p / num_trials

In [19]:
print(estimate_birthday_probability(10000,23))

0.5025


## Part 2: Diagnostic tests and Bayes rule ( pts)

Let's return to the case of finding lead pipes discussed in class. Fully excavating pipes leading from the street to a house is expensive, so contractors in Flint instead used a simpler method that exposed only a part of the pipes.  Despite being cheaper than a full excavaction, one issue with this method is that it risks missing parts of the pipe containing lead. Your job is to decide whether this is a good test.

To begin, symbolically write the probabilities $P(lead|positive)$ and $P(lead|negative)$ in terms of the prior probability that a pipe contains led and the probabilities of test outcome conditional on pipe contamination using Bayes rule. (This is a bit messy to do in a notebook with your current knowledge, so write it on paper first then try your best to transcribe it by typing).

Now that we've done that, write a function that takes three arguments:
- The prior probability of lead $P(lead)$
- The probability of a positive test if lead is present $P(positive|lead)$
- The probability of a positive test if lead is *not* present $P(positive|not lead)$. 

It should return a numpy array containing $P(lead|positive)$ and $P(lead|negative)$.

Think about your variable names, this is one of (many) cases that we'll encounter where quantities might have similar names and we need to be careful to not confuse them.

Now that we've done that, write a function that takes three arguments:
- The prior probability of lead $P(lead)$
- The probability of a positive test if lead is present $P(positive|lead)$
- The probability of a positive test if lead is *not* present $P(positive|not lead)$. 

It should return a numpy array containing $P(lead|positive)$ and $P(lead|negative)$.

Think about your variable names, this is one of (many) cases that we'll encounter where quantities might have similar names and we need to be careful to not confuse them.

In [15]:
##answer here
def compute_prob(p_lead, p_positivegivenlead,p_positivegivennNOlead):
 p_NOlead = 1 - p_lead
 positive = p_positivegivenlead*p_lead + p_positivegivenNOlead*p_NOlead
 leadgivenpos = (p_positivegivenlead*p_lead) / (positive)

 p_negativegivenlead = 1 - p_positivegivenlead
 p_negativegivenNOlead = 1 - p_positivegivenNOlead
 negative = p_negativegivenlead * p_lead + p_negativegivenNOlead * p_NOlead
 p_leadgivenneg = p_negativegivenlead*p_lead / negative

 array = [leadgivenpos,p_leadgivenneg]
 return array

In [16]:
p_lead = 0.85
p_positivegivenlead = 0.9
p_positivegivenNOlead = 0.1

print(compute_prob(p_lead, p_positivegivenlead,p_positivegivenNOlead))


[0.9807692307692307, 0.3863636363636362]


Suppose you live in a neighborhood of Flint which has older homes built with low-quality materials near the city center, and suppose that the prior probability of such a home having lead pipes is 0.85. Assume further that the test is 90% accurate: if there is lead the test will be positive with probability 0.9, and if there is no lead the test will be positive with probability 0.1.

What is the output of your function for these probabilities?

##answer here : [0.9807692307692307, 0.019230769230769273]

You just received notification that according to the 90% accurate test you do not have lead. Do you let your children drink your tapwater? Why or why not?

Answer here: No, I wouldn't let them drink the water if there is still a 10% chance that the test is wrong and it does contain lead.

### Part 3: Choose your weapon ( pts)

In the game Dungeons & Dragons we simulate combat by rolling dice. These come in multiple shapes, from four-sided (d4) to 20-sided (d20). Each type of weapon has a number and shape of dice to determine the damage inflicted by a successful blow. Consider two weapons: a 1d8 (roll one eight-sided die) and a 2d4 (roll two four-sided dice).

Calculate the expected damage from each weapon by following the below steps. 

1. For each weapon use one or more for loops to create an array containing all possible outcomes of the die rolls for that weapon (including duplicates if any). 
(For 1d6 this would be `[1,2,3,4,5,6]` for 2d6 `[(1,1), (1,2), ...., (6,6)]`)
2. Create a `Counter` from each array.
3. Iterate over the elements of this `Counter` in a for loop. 

   3a. For each value calculate the probability of that value (using the number of times it occurs in the `Counter`) and the value times its probability. 
   
   3b. Print them (the value, its probability, and their product). 
4. After that loop, display the sum of these value/probability products, which is the expectation of the damage.

In [4]:
##your answer here
from collections import Counter

# roll 1 eight-sided die
eightsided = []
for x in range(1,9):
    eightsided.append(x)
    print(eightsided[x])   
print(counter1)
summer1 = []
supersum1 = 0
for i in range(0,16):
    sum1 = counter1.count(i+1)
    prob1 = counter1.count(i+1) / 8
    summer1.append((i+1)*prob1)
    print(i+1, ' : ',sum1, prob1, summer1[i])
    supersum1 = supersum1 + summer1[i]
print(supersum1)

# roll 2 four-sided dice
foursided = []
counter2 = []
for x in range(1,5):
    x1 = x
    for y in range(1,5):
        y1 = y
        b = (x1,y1)
        foursided.append(b)
        counter2.append(x + y + 2) 
for x in range(0,16):
    print(foursided[x])
print(counter2)
summer = []
supersum = 0
for i in range(0,16):
    sum = counter2.count(i+1)
    prob = counter2.count(i+1) / 16
    summer.append((i)*prob)
    print(i+1, ' : ',sum, prob, summer[i])
    supersum = supersum + summer[i]
print(supersum)
        

IndexError: list index out of range

Now calculate the variance of the weapons, **using the x
list comprehension syntax**. 

Starting with the lists of possible outcomes (including any duplicates) that you created in the previous block, create new lists containing the square of the difference between each value and the expectation that you calculated for that weapon. 

As an example, if my data points were 1, 2, 4 then the expected value would be $\bar{x}=\frac{7}{3}$ and the list would be $[(1-\bar{x})^2, (2-\bar{x})^2, (4-\bar{x})^2]$

Print these new lists and their respective averages, which are the variances of the distributions.

In [71]:
##answer here

# 8 sided die
expected_values2 = []
nowsum = 0
for i in range(0,len(counter1)):
    x = (counter1[i]-4.5)**2 
    expected_values2.append(x)
    nowsum += x
print(expected_values2)
avg1 = nowsum / 8
print(avg1)

# 4 sided die
expected_values = []
nowsum2 = 0
for i in range(0,len(counter2)):
    x = (counter2[i]-4.0)**2 
    expected_values.append(x)
    nowsum2 += x
print(expected_values)
avg2 = nowsum2 / 16
print(avg2)

[12.25, 6.25, 2.25, 0.25, 0.25, 2.25, 6.25, 12.25]
5.25
[4.0, 1.0, 0.0, 1.0, 1.0, 0.0, 1.0, 4.0, 0.0, 1.0, 4.0, 9.0, 1.0, 4.0, 9.0, 16.0]
3.5


Which is better? Describe why you might prefer the 1d8 weapon or the 2d4 weapon. (No game knowledge is necessary, we will grade based on your reasoning.)

Answer here: The 1d8 weapon is more powerful as it yeilds a greater average damage, however it is not as reliable as it has a greater variance. On the other hand the 2d4 weapon yields a lower average damage but is more reliable as it has a lower variance. Generally the 1d8 would be a better but slightly riskier choice. 